instrumentation/armeria-1.0/auto/src/main/java/io/opentelemetry/auto/instrumentation/armeria/v1_0/ArmeriaServerBuilderInstrumentation.java

```diff
@@ -0,0 +1,84 @@
+ * Copyright The OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.opentelemetry.auto.instrumentation.armeria.v1_0;
+
+import static net.bytebuddy.matcher.ElementMatchers.isConstructor;
+import static net.bytebuddy.matcher.ElementMatchers.isMethod;
+import static net.bytebuddy.matcher.ElementMatchers.isPublic;
+import static net.bytebuddy.matcher.ElementMatchers.named;
+import static net.bytebuddy.matcher.ElementMatchers.takesArgument;
+
+import com.google.auto.service.AutoService;
+import com.linecorp.armeria.server.ServerBuilder;
+import io.opentelemetry.auto.instrumentation.armeria.v1_0.shaded.server.OpenTelemetryService;
+import io.opentelemetry.auto.tooling.Instrumenter;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.function.Function;
+import net.bytebuddy.asm.Advice;
+import net.bytebuddy.description.method.MethodDescription;
+import net.bytebuddy.description.type.TypeDescription;
+import net.bytebuddy.matcher.ElementMatcher;
+
+@AutoService(Instrumenter.class)
+public class ArmeriaServerBuilderInstrumentation extends AbstractArmeriaInstrumentation {
+
+  @Override
+  public ElementMatcher<? super TypeDescription> typeMatcher() {
+    return named("com.linecorp.armeria.server.ServerBuilder");
+  }
+
+  @Override
+  public Map<? extends ElementMatcher<? super MethodDescription>, String> transformers() {
+    Map<ElementMatcher<MethodDescription>, String> transformers = new HashMap<>();
+    transformers.put(
+        isConstructor(),
+        ArmeriaServerBuilderInstrumentation.class.getName() + "$ConstructorAdvice");
+    transformers.put(
+        isMethod().and(isPublic()).and(named("decorator").and(takesArgument(0, Function.class))),
+        ArmeriaServerBuilderInstrumentation.class.getName() + "$SuppressDecoratorAdvice");
+    return transformers;
+  }
+
+  public static class ConstructorAdvice {
+    @Advice.OnMethodExit
+    public static void construct(@Advice.This ServerBuilder builder) {
+      builder.decorator(OpenTelemetryService.newDecorator());
+    }
+  }
+
+  public static class SuppressDecoratorAdvice {
+    @Advice.OnMethodEnter(skipOn = Advice.OnNonDefaultValue.class)
+    public static boolean suppressDecorator(@Advice.Argument(0) Function<?, ?> decorator) {
+      return decorator
+          .getClass()
+          .getName()
+          .equals(
+              "io.opentelemetry.instrumentation.armeria.v1_0.server.OpenTelemetryService$Decorator");
```

Comment:
i think #892 will break this because it will shade this constant also. there are ways to prevent the string from getting shaded (e.g. make it a constant NAME = "_io.opentelemetry...".substring(1)".

another option for suppressing library instrumentation that might be easier to generalize, is to make the coordination more explicit, e.g. have a method in `instrumentation-api` like `isAutoEnabled(instrumentationName)`, and the library instrumentation can explicitly check that to see if there is auto instrumentation with the same name enabled

in this case i guess that would mean returning a no-op decorator, or changing the API to expose a method that registers a decorator on the `ServerBuilder` so we can no-op the registration in this case

I'll add this note to #903, no need to decide/address in this PR

Reply:
I ended up needing yet a different pattern because of the shading change making the classnames the same. I guess it's actually a bit more robust now and we'll revisit in #903 for an inversion of control.